packages/core/src/components/overlay/overlay.tsx

```diff
@@ -425,7 +425,7 @@ export class Overlay extends AbstractPureComponent2<IOverlayProps, IOverlayState
 
     private handleDocumentClick = (e: MouseEvent) => {
         const { canOutsideClickClose, isOpen, onClose } = this.props;
-        const eventTarget = e.target as HTMLElement;
+        const eventTarget = (e.composed ? e.composedPath()[0] : e.target) as HTMLElement;
```

Comment:
is `e.composedPath()` always guaranteed to be non-empty? can you add some code comments to explain why we are adding this special logic here and on L446?

Reply:
> is `e.composedPath()` always guaranteed to be non-empty?

Yes. It returns _an array of EventTarget objects representing the objects on which an event listener will be invoked (from [MDN](https://developer.mozilla.org/en-US/docs/Web/API/Event/composedPath#Return_value)), the first item is always equals to the `event.target`, or if this event is triggered in an open mode Shadow DOM, the first item is the actually element that trigger the event. There are [some examples in MDN](https://developer.mozilla.org/en-US/docs/Web/API/Event/composedPath#Examples).

>  can you add some code comments to explain why we are adding this special logic here and on L446?

Done. A more detailed explanation can be found in #4220